### PR TITLE
Makes ng-gauge highlights width configurable

### PR DESCRIPTION
**Pull request by jaredjensen**
This adds a "highlights width" property to the radial gauge config with a range of `0-25`.  It still defaults to `5`, but that value is too skinny for my old eyes.

image
image

It also looks like there's a bug that prevents zones (and therefore highlights) from being available on initial load.  I clumsily fixed this in the 2nd commit by moving `initWidget()` after `observeDataStream()`, otherwise it seems the data service `_pathRegister` isn't loaded when the gauge tries to subscribe to metadata.  There's probably a better way to solve this, but it felt that would require a more fundamental change that would affect all widgets (although other widgets likely have this problem, too).

### CODE REVIEW

**Comment by godind** [comment]
Thanks for this PR. I feel bad for sharing this but it's worth it.

In the next version (v3), gauges progress bars and gauge value change colors as they enter/leave zones. KIP will also have an all black theme with a set of 7 bright widget colors to help with sunlight visibility. I'll add an all white next which, even if I prefer black, look sharper in direct sunlight.

Maybe you can take a look at the beta and see if this is eliminates the need to resize highlights? It's an unfinished beta product meant to test the new "gesture first" layout (no bottom navbar, swipe in four directions to obtain the same navbar fonctions.

**Comment by jaredjensen** [comment]
@godind I did see the comment about improving gauges in v3 in https://github.com/mxtommy/Kip/issues/410, and the improvements you described sound great!  I started building a "green is good" style native linux app to achieve the same thing when I stumbled on your plugin.

I'm happy to test v3, but I made this PR because we had an alternator fire on our last trip, and I'm feeling a bit urgent about getting better monitoring in place.  I can run my fork until v3 is available if you don't want to merge it, though.

**Comment by godind** [comment]
I'll push it in. Can you also do the linear and steelgauge please?

**Comment by jaredjensen** [comment]
You bet, I'll try to knock those out tonight.

**Comment by godind** [comment]
Thanks! If you want to try the beta:
1- crate another SK user
2- sign in with that user so that you don't mess up your existing config.
3- run npm I @mxtommy/kip@beta from the ~/.signalk folder.
4- restart SK and launch the server. You can load the demo config to get started.

To uninstall, use the SK AppStore UI uninstall and reinstall the latest.

It's a work in progress so not close to final. Have a look at the gauge zones and colour. It should be more obvious now. Use swipe up/down to change dashboard and swipe left/right to open action and notification sidebars.

Have fun.

**Comment by jaredjensen** [comment]
I pushed a couple more commits to add highlights width to the linear gauge and fix the initial load issue with the steel gauge.  I don't think I can make the highlights width configurable on the steel gauge because [it seems to be hard-coded](https://github.com/nicolas-van/steelseries/blob/master/src/Radial.js#L573), if I'm reading it right.

**Comment by godind** [comment]
Oh. True. Steelserie gauge color the existing bar. Sorry I forgot. 

I'll check the PR today or tomorrow as I have a few days to play with KIP.

File: src/app/widget-config/modal-widget-config/modal-widget-config.component.html [hunk 1]
```diff
@@ -168,7 +167,10 @@
                       <mat-option value="capacity">Capacity</mat-option>
                     </mat-select>
                   </mat-form-field>
-
+                  <mat-form-field formGroupName="gauge" class="options-grid-span2">
```
Comment (godind) [inline src/app/widget-config/modal-widget-config/modal-widget-config.component.html:170]: Would'nt we be better of with a single new @if statement rather then repeating the template code?

File: src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts [hunk 2]
```diff
@@ -128,6 +128,8 @@ export class WidgetGaugeNgRadialComponent extends BaseWidgetComponent implements
       }
     });
 
+    this.initWidget();
```
Comment (godind) [inline src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts:131]: same here

File: src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts [hunk 3]
```diff
@@ -171,7 +173,7 @@ export class WidgetGaugeNgRadialComponent extends BaseWidgetComponent implements
     this.gaugeOptions.valueDec = this.widgetProperties.config.numDecimal !== undefined && this.widgetProperties.config.numDecimal !== null ? this.widgetProperties.config.numDecimal : 2;
     this.gaugeOptions.majorTicksInt = this.widgetProperties.config.numInt !== undefined && this.widgetProperties.config.numInt !== null ? this.widgetProperties.config.numInt : 1;
     this.gaugeOptions.majorTicksDec = this.widgetProperties.config.numDecimal !== undefined && this.widgetProperties.config.numDecimal !== null ? this.widgetProperties.config.numDecimal : 2;
-    this.gaugeOptions.highlightsWidth = 0;
+    this.gaugeOptions.highlightsWidth = this.widgetProperties.config.gauge.highlightsWidth !== undefined && this.widgetProperties.config.gauge.highlightsWidth !== null ? this.widgetProperties.config.gauge.highlightsWidth : 5;
```
Comment (godind) [inline src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts:176]: Reworking initWidget() will remove the need to test highlightsWidth for undefined or null.

File: src/app/widgets/widget-gauge-steel/widget-gauge-steel.component.ts [hunk 4]
```diff
@@ -72,6 +71,8 @@ export class WidgetGaugeComponent extends BaseWidgetComponent implements OnInit,
       }
     );
 
+    this.initWidget();
```
Comment (godind) [inline src/app/widgets/widget-gauge-steel/widget-gauge-steel.component.ts:74]: see previous comment

File: src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts [hunk 5]
```diff
@@ -252,7 +254,7 @@ export class WidgetGaugeNgLinearComponent extends BaseWidgetComponent implements
       needleSide: "both",
 
       highlights: [],
-      highlightsWidth: 0,
+      highlightsWidth: this.widgetProperties.config.gauge.highlightsWidth !== undefined && this.widgetProperties.config.gauge.highlightsWidth !== null ? this.widgetProperties.config.gauge.highlightsWidth : 5,
```
Comment (godind) [inline src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts:257]: Reworking initWidget() will remove the need to test highlightsWidth for undefined or null.

**Review by godind — CHANGES_REQUESTED** [review]
See inline comments for review.

File: src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts [hunk 6]
```diff
@@ -130,6 +130,8 @@ export class WidgetGaugeNgLinearComponent extends BaseWidgetComponent implements
       }
     });
 
+    this.initWidget();
```
[the remark this answers is not in the document]
Comment (jaredjensen) [reply src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts:133]: That's essentially what I discovered when I started trying to understand why the zones weren't displaying consistently.  I put that change in a separate commit and noted it was "clumsy" in the PR description, basically trying to highlight the issue and how naive resequencing fixes it, but also likely causes a bunch of other problems.

Given you're working on v3, I'm happy to revert that commit and just know that I'll have to switch to a different page (e.g. settings) and back to my gauges once at startup for the zones to show.

File: src/app/widget-config/modal-widget-config/modal-widget-config.component.html (hunk 1 above)
Comment (godind) [reply src/app/widget-config/modal-widget-config/modal-widget-config.component.html:170]: The component uses formbuilder to dynamically parse the config object.

The template part is like 7 years old and ugly - ageed! The logic is just to run by the config or form object properties and values and output the approriate field controls. It's all fits in a 2 columns grid layout. It's not super flexible from a UX prespective, but with lots of Widgets and options, it needed something generic. You can use @if and @switch to adjust say, column span base on config properties, if necessary. I'm afraid that by repeating it this way, we will forget to check all config options and conditionnal cases. Months can go by until coding work and we tend to forget these things...plus we have de tests!

I'm all ears to see someone submit a PR with something better.

File: src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (godind) [reply src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts:133]: If you can spend some time fixing it. Revert it from the current PR so we have the highlightWidth stuff and push another PR just with the Zones init fix. I think it's just a matter of moving observeMeta into observeDatastream in base-widget, but it requires some good manual testing.

Ny the way, I really do appreciate the help!

File: src/app/widget-config/modal-widget-config/modal-widget-config.component.html (hunk 1 above)
Comment (jaredjensen) [reply src/app/widget-config/modal-widget-config/modal-widget-config.component.html]: Totally get it.  I've been doing this 30+ years and have my own experience with maintaining "legacy" code, often of my own creation. =)

File: src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaredjensen) [reply src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts]: I'll give that a go in another branch.  I think I've addressed your feedback in this PR, but let me know if I've missed something.

File: src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts (hunk 3 above)
Comment (jaredjensen) [reply src/app/widgets/widget-gauge-ng-radial/widget-gauge-ng-radial.component.ts]: I think specifically adding `highlightsWidth` to `defaultConfig` makes this check unnecessary.  I tested that by manually editing the local storage config to remove `highlightsWidth` and reloading the page, and the result was that the default width was used.

File: src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaredjensen) [reply src/app/widgets/widget-gauge-ng-linear/widget-gauge-ng-linear.component.ts]: The change to `initWidget()` to fix the metadata issue does appear to work.  I put that change in https://github.com/jaredjensen/Kip/pull/1 and we can continue the conversation there.  Once you merge this PR (for highlights width), I'll rebase `zone-init-fix` and update that PR.

**Comment by godind** [comment]
@jaredjensen merged. Thanks!
